tools/crm_mon.c

```diff
@@ -2807,18 +2853,363 @@ print_failed_actions(FILE *stream, pe_working_set_t *data_set)
+                    }
+                    np = np->next;
+                } while (1);
+            }
+        } else {
+            /* put into pending */
+            hp->next = pending;
+            pending = hp;
+        }
+        hp = tmp;
+    }
+
+    /* pending actions don't have a completed-stamp so make them go front */
+    if (pending) {
+        stonith_history_t *last_pending = pending;
+
+        while (last_pending->next) {
+            last_pending = last_pending->next;
+        }
+
+        last_pending->next = new;
+        new = pending;
+    }
+    return new;
+}
+
+/*!
+ * \internal
+ * \brief Turn stonith action into a better readable string
+ *
+ * \param[in] action     Stonith action
+ */
+static char *
+fence_action_str(const char *action)
+{
+    char *str = NULL;
+
+    if (action == NULL) {
+        str = strdup("unknown");
+    } else if (action[0] == 'o') { // on, off
+        str = crm_concat("turn", action, ' ');
+    } else {
+        str = strdup(action);
+    }
+    return str;
+}
+
+/*!
+ * \internal
+ * \brief Print a stonith action
+ *
+ * \param[in] stream     File stream to display output to
+ * \param[in] event      stonith event
+ */
+static void
+print_stonith_action(FILE *stream, stonith_history_t *event)
+{
+    char *action_s = fence_action_str(event->action);
+    char *run_at_s = ctime(&event->completed);
+
+    if ((run_at_s) && (run_at_s[0] != 0)) {
+        run_at_s[strlen(run_at_s)-1] = 0; /* Overwrite the newline */
+    }
+
+    switch(output_format) {
+        case mon_output_xml:
+            fprintf(stream, "        <fence_event target=\"%s\" action=\"%s\"",
+                    event->target, event->action);
+            switch(event->state) {
+                case st_done:
+                    fprintf(stream, " state=\"success\"");
+                    break;
+                case st_failed:
+                    fprintf(stream, " state=\"failed\"");
+                    break;
+                default:
+                    fprintf(stream, " state=\"pending\"");
+            }
+            fprintf(stream, " origin=\"%s\" client=\"%s\"",
+                    event->origin, event->client);
+            if (event->delegate) {
+                fprintf(stream, " delegate=\"%s\"", event->delegate);
+            }
+            switch(event->state) {
+                case st_done:
+                case st_failed:
+                    fprintf(stream, " completed=\"%s\"", run_at_s?run_at_s:"");
+                    break;
+                default:
+                    fprintf(stream, " state=\"pending\"");
+            }
+            fprintf(stream, " />\n");
+            break;
+
         case mon_output_plain:
         case mon_output_console:
-            print_as("\n");
+            switch(event->state) {
+                case st_done:
+                    print_as("* %s of %s successful: delegate=%s, client=%s, origin=%s,\n"
+                             "    %s='%s'\n",
```

Comment:
Let's not take the Failed Actions list as an example of user-friendly output :)

Let's do something readable, maybe something like:
`<action> of <target> by <delegate> for <client> on <origin> <succeeded|failed> at <time>`